transport/ice/IceSession.cpp

```diff
@@ -472,11 +493,252 @@ bool IceSession::isResponseAuthentic(const void* data, size_t len) const
         stunMessage->isAuthentic(_hmacComputer.remote);
 }
 
+bool IceSession::isIceAuthentic(const void* data, size_t len) const
+{
+    if (ice::isResponse(data))
+    {
+        return isResponseAuthentic(data, len);
+    }
+
+    if (ice::isRequest(data))
+    {
+        return isRequestAuthentic(data, len);
+    }
+
+    return false;
+}
+
+bool IceSession::processValidStunTcpRequest(IceEndpoint* localEndpoint,
+    const transport::SocketAddress& remotePort,
+    const StunMessage& msg,
+    int remoteCandidatePriority,
+    uint64_t now)
+{
+    assert(localEndpoint->getTransportType() == TransportType::TCP);
+
+    if (isAttached(localEndpoint))
+    {
+        sendResponse(localEndpoint, remotePort, 0, msg, now);
+        return true;
+    }
+
+    IceCandidate remoteCandidate(_component,
+        localEndpoint->getTransportType(),
+        remoteCandidatePriority,
+        remotePort,
+        remotePort,
+        IceCandidate::Type::PRFLX);
+
+    auto* candidatePair = addRemoteTcpCandidateAndCreatePair(remoteCandidate, localEndpoint);
+    if (candidatePair)
+    {
+        candidatePair->accept();
+        sendResponse(localEndpoint, remotePort, 0, msg, now);
+        candidatePair->receptionTimestamp = now;
+        candidatePair->send(now);
+    }
+
+    if (_state == State::CONNECTING)
+    {
+        sortCheckList();
+    }
+
+    return candidatePair != nullptr;
+}
+
+bool IceSession::processValidStunUdpRequest(IceEndpoint* localEndpoint,
+    const transport::SocketAddress& remotePort,
+    const StunMessage& msg,
+    int remoteCandidatePriority,
+    uint64_t now)
+{
+    assert(localEndpoint->getTransportType() == TransportType::UDP);
+
+    if (!(_state == State::CONNECTING || _state == State::CONNECTED))
+    {
+        return false;
+    }
+
+    auto* pair = findCandidatePair(localEndpoint, remotePort);
+    if (pair)
+    {
+        // Only send response after accept! Otherwise we could drop early media
+        pair->accept();
```

Comment:
first accept pair. This will call the listener to register the endpoint (when it is not "accepted" already). Only then send response to be sure clients considers this a valid pair when we have already register the peer on udp endpoint